docs/build/dapp-staking/dapp_staking_architecture.md

```diff
@@ -0,0 +1,672 @@
+Rewards are ONLY issued when they are being claimed.
+
+## II. Public Calls & Internals
+
+These are the name of the calls exposed by the pallet. Whatever naming is used in frontend can be entirely different.
+
+### 1. Internals
+
+This covers important internal calls which are critical part of the dApps staking v3 design.
+
+#### >> on_initialize
+Executed at the beginning of each block. The solution can be easily extended to utilize scheduler pallet later instead of `on_initialize`.
+
+Checks if new era or new period needs to begin:
+- In case no new era/periods should start from this block, do nothing;
+- If pallet is in maintenance mode, do nothing;
+- In case changes are needed:
+    - **New era during build&earn period:**
+        - calculate dApp tier allocation & reward pool;
+     - assign staker reward pool;
+    - **Final era of build&earn period:**
+        - switch over to the voting period;
+        - this also means switching over to the next period;
+        - calculate new tier configuration;
+    - **New era during voting period:**
+        - switch over to the build&earn period;
+
+:::note
+
+Storage cleanup should occur after switching over to a new period, or can be done lazily via hooks.
+
+:::
+
+#### >> get_dapp_tier_assignment
+
+Iterates over all dApps and their stake, and distributes them into tiers. A simple way of tier assignment is used, no complex formula is involved - **total stake amount**.  
+dApps are assigned into tiers, and reward per tier is calculated.
+
+:::note
+There are two features missing from the current implementation:  
+Handling for a corner case when multiple dApps have the same score, but tier hasn’t got enough capacity.
+:::
+
+The calculation is heavy on the DB reads, and includes two sorts.
+
+### 2. Register
+
+`register(developer, contract)` : Registers a new developer *account < > smart contract* pair. 
+
+It's a permissioned action, origin defined by a configurable type. *(In case normal user calls `register`, an error is thrown)*  
+There is no registration deposit.
+
+Even though each dApp has a unique address, an additional unique `u16` dApp Id is assigned to the dApp once it’s been created.
+
+### 3. Unregister
+
+`unregister(contract)` : Unregisters the contract, marking it as unregistered from current era. 
+
+Removes storage entry (or entries) related to that contract, except for info when it was unregistered.
+
+### 4. Lock
+
+`lock(amount)` : Locks up some amount in dApps staking.
+
+*Frontend will ensure user never locks their entire balance in dApp staking, preventing them from paying fees in the future.* 
+
+It’s enough to store how much user has locked **right now**, there's no need for historical data.
+Locked amount should always be above some threshold to prevent account spamming (inflating number of dApp staking participants).
+
+In case user specifies value which is higher than what is available for locking, lock what’s available.
```

Comment:
This also needs to be re-worded to something like:

`In case user specifies value which is higher than what's available for locking, the maximum available amount will be locked.`.